### PR TITLE
Improved performance of django.forms.ChoiceWidget.optgroups()

### PR DESCRIPTION
**Pull request by smithdc1**
I was looking at SelectDateWidget as I thought I spotted an ideal candidate for a list comprehension. When I looked into it I found that the slowest part of this benchmark is creating all of the options, so I investigated this instead.

I found two items that could help here. The first is re-ordering the select logic, the `str` and `in` are more expensive than an attribute lookup. 

The second change is to "cache" the `create_option` function (I'm not sure what the correct way of describing this is). I'm less sure about this change, but it _does_ have a sizeable impact so I've suggested it. 

The only other thing I considered here was also "caching" `self.allow_multiple_selected` as it _should_ be quicker to store that as a local variable. However, I can't detect the impact of this change, so it's not proposed here. 

Finally, this list is _quite_ long and so performance gains are larger as there are more loops. I tested with a shorter, simpler widget and I couldn't detect any impact of this change. Therefore I think it benefits larger option lists and doesn't detriment smaller ones.

Start
`SelectDateWidget: Mean +- std dev: 4.50 ms +- 0.14 ms`

re-order logic (4%)
`SelectDateWidget: Mean +- std dev: 4.31 ms +- 0.19 ms`

_cache_ create_option (a further 4%; 8% total)
`SelectDateWidget: Mean +- std dev: 4.13 ms +- 0.11 ms`

<details>
<summary>bench_selectdatewidget.py</summary>
<p>

``` python
from django.conf.global_settings import INSTALLED_APPS
import django
from django.conf import settings
from django.forms.widgets import SelectDateWidget
import pyperf

settings.configure(INSTALLED_APPS= ('__main__',))
django.setup()


def test(loops):
    range_it = range(loops)
    widget = SelectDateWidget(years=(2020,))
    t0 = pyperf.perf_counter()

    # repeat to reduce impact of for loop
    for loop in range_it:
        widget.get_context('widget', "2020-10-10", {})
        widget.get_context('widget', "2020-10-10", {})
        widget.get_context('widget', "2020-10-10", {})
        widget.get_context('widget', "2020-10-10", {})
        widget.get_context('widget', "2020-10-10", {})
        widget.get_context('widget', "2020-10-10", {})
        widget.get_context('widget', "2020-10-10", {})
        widget.get_context('widget', "2020-10-10", {})
        widget.get_context('widget', "2020-10-10", {})
        widget.get_context('widget', "2020-10-10", {})

    return pyperf.perf_counter() - t0

runner = pyperf.Runner()
runner.bench_time_func('SelectDateWidget', test)

```

</p>
</details>

### CODE REVIEW

File: django/forms/widgets.py [hunk 1]
```diff
@@ -588,6 +588,7 @@ def optgroups(self, name, value, attrs=None):
         """Return a list of optgroups for this widget."""
         groups = []
         has_selected = False
+        create_option = self.create_option
```
Comment (carltongibson) [inline django/forms/widgets.py:591]: I'm sceptical about this. Generating options is slow. I'm not convinced that avoiding the attribute lookup is the way to go with it though. A 4% speed-up is OK but the cost is obfuscate the code.   

We used to do this with JavaScript. Browsers were bad at navigating the scope hierarchy, so you'd create local copies of variables at the start of a function. It worked, a good speedup. But it was horrible. In the end they made browsers better, and I don't see it done these days (certainly not as a first-option). Ultimately it was the browser's issue. 

Same here. If this is an issue, it's Python's issue (IMO) — let them speed up attribute access. (If we're that worried about speed, there's a question about why we're using Python in the first place... but that gets us off-topic). 

`SelectDateWidget` is close to a worst-case example. I'm not convinced it's typical. Is it really the case that choices generation is the holdup in common scenarios…?

**And** if you are hitting issues here, the way forward (again IMO) is to cache choices generation, rather than bend the code the wrong way. I see us removing old micro-optimizations as unnecessary; not at all sure about adding more...

**Review by carltongibson — COMMENTED** [review]
Hey @smithdc1 — I'm not sure about this. 

Seems we've come down against a few of these, so perhaps we should discuss (although we have merged others so...) 

tl;dr: I think we should take the one, but not the other. What do you think? 

Thanks!

**Comment by smithdc1** [comment]
I guess part of this is myself learning what is and isn't acceptable. I feel judging between various ways of doing things comes with experience. 

I guess my thoughts are in part shown by me splitting the change. The re-ordering of the logic seems sensible to me. We're exploiting pythons ability to fail fast and doesn't impact code readability. The second I was less sure about, reading around I see folk doing this but does come at a cost as you describe. 

A good outcome for me here is to learn what is and isn't appropriate.

**Comment by carltongibson** [comment]
Hey @smithdc1, thanks. I think ultimately readability trumps unless we can show that something is on the critical path in realistic circumstances. (Something in URL resolution, for example 😃, is worth leaning more towards speed.) 

If you want to reduce this down to just the Boolean change, we’d be happy with that. 

👍

**Comment by smithdc1** [comment]
Thank you both! 🙏
